wire/protobuf/serializer.go

```diff
@@ -0,0 +1,145 @@
+
+import (
+	"encoding/binary"
+	"io"
+	"time"
+
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/proto"
+	"perun.network/go-perun/wire"
+)
+
+// Serializer implements methods for encoding/decoding
+// envelopes using protobuf encoding.
+type Serializer struct{}
+
+// func init() {
+// 	wire.SetEnvelopeSerializer(Serializer{})
+// }
+
+// Encode encodes the envelope into the wire using perunio
+// encoding format.
+func (Serializer) Encode(w io.Writer, env *wire.Envelope) error {
+	sender, err := env.Sender.MarshalBinary()
+	if err != nil {
+		return errors.WithMessage(err, "marshalling sender address")
+	}
+	recipient, err := env.Recipient.MarshalBinary()
+	if err != nil {
+		return errors.WithMessage(err, "marshalling recipient address")
+	}
+
+	var grpcMsg isEnvelope_Msg
+	switch env.Msg.Type() {
+	case wire.Ping:
+		msg, ok := env.Msg.(*wire.PingMsg)
+		if !ok {
+			return errors.New("message type and content mismatch")
+		}
+		grpcMsg = &Envelope_PingMsg{
+			PingMsg: &PingMsg{
+				Created: msg.Created.UnixNano(),
+			},
+		}
+	case wire.Pong:
+		msg, ok := env.Msg.(*wire.PongMsg)
+		if !ok {
+			return errors.New("message type and content mismatch")
+		}
+		grpcMsg = &Envelope_PongMsg{
+			PongMsg: &PongMsg{
+				Created: msg.Created.UnixNano(),
+			},
+		}
+	case wire.Shutdown:
+		msg, ok := env.Msg.(*wire.ShutdownMsg)
+		if !ok {
+			return errors.New("message type and content mismatch")
+		}
+		grpcMsg = &Envelope_ShutdownMsg{
+			ShutdownMsg: &ShutdownMsg{
+				Reason: msg.Reason,
+			},
+		}
+	}
+
+	protoEnv := Envelope{
+		Sender:    sender,
+		Recipient: recipient,
+		Msg:       grpcMsg,
+	}
+
+	data, err := proto.Marshal(&protoEnv)
+	if err != nil {
+		return errors.Wrap(err, "marshalling envelope")
+	}
+
+	if err := binary.Write(w, binary.BigEndian, uint16(len(data))); err != nil {
+		return errors.Wrap(err, "writing length to wire")
+	}
+
+	_, err = w.Write(data)
+	return errors.Wrap(err, "writing data to wire")
+}
+
+// Decode decodes an envelope from the wire using perunio encoding format.
+func (Serializer) Decode(r io.Reader) (*wire.Envelope, error) {
+	var length uint16
+	if err := binary.Read(r, binary.BigEndian, &length); err != nil {
+		return nil, errors.Wrap(err, "reading length from wire")
+	}
+	data := make([]byte, length)
+	if _, err := r.Read(data); err != nil {
+		return nil, errors.Wrap(err, "reading data from wire")
+	}
+	var protoEnv Envelope
+	if err := proto.Unmarshal(data, &protoEnv); err != nil {
+		return nil, errors.Wrap(err, "unmarshalling envelope")
+	}
+
+	env := wire.Envelope{}
+	env.Sender = wire.NewAddress()
+	if err := env.Sender.UnmarshalBinary(protoEnv.Sender); err != nil {
+		return nil, errors.Wrap(err, "unmarshalling sender address")
+	}
+	env.Recipient = wire.NewAddress()
+	if err := env.Recipient.UnmarshalBinary(protoEnv.Recipient); err != nil {
+		return nil, errors.Wrap(err, "unmarshalling recipient address")
+	}
+
+	switch protoEnv.Msg.(type) {
+	case *Envelope_PingMsg:
```

Comment:
I realize that this PR is not complete yet as many message types are missing, right?

Reply:
Yes.